src/ASM/DomainDecomposition.C

```diff
@@ -0,0 +1,787 @@
+      auto dof = sim.getSAM()->getNodeDOFs(i);
+      maxDof += dof.second-dof.first+1;
+      for (auto& it : MLGN)
+        utl::renumber(it, old2new2, false);
+    }
+
+  if (adm.getProcId() < adm.getNoProcs()-1) {
+    adm.send(maxNode, adm.getProcId()+1);
+    adm.send(maxDof, adm.getProcId()+1);
+  }
+
+  for (const auto& it : ghostConnections) {
+    int midx = sim.getLocalPatchIndex(it.master);
+    if (midx < 1)
+      continue;
+
+    std::vector<int> glbNodes;
+    sim.getPatch(midx)->getBoundaryNodes(it.midx, glbNodes);
+    for (size_t i = 0; i < glbNodes.size(); ++i)
+      glbNodes[i] = MLGN[glbNodes[i]-1];
+
+    adm.send(int(glbNodes.size()), getPatchOwner(it.slave));
+    adm.send(glbNodes, getPatchOwner(it.slave));
+  }
+
+#endif
+
+  return true;
+}
+
+
+std::vector<int> DomainDecomposition::setupEquationNumbers(const SIMbase& sim,
+                                                           int pidx, int lidx)
+{
+  std::vector<IntVec> lNodes(sim.getPatch(pidx)->getNoBasis());
+  std::vector<int> result;
+
+  for (size_t block = 0; block < blocks.size(); ++block) {
+    std::set<int> bases;
+    if (block == 0) {
+      for (size_t i = 1; i <= sim.getPatch(pidx)->getNoBasis(); ++i)
+        bases.insert(i);
+    } else
+      bases = utl::getDigits(sim.getSolParams()->getBlock(block-1).basis);
+
+    for (const int& basis : bases) {
+      if (lNodes[basis-1].empty())
+        sim.getPatch(pidx)->getBoundaryNodes(lidx, lNodes[basis-1], basis);
+
+      std::set<int> components;
+      if (block == 0 || sim.getSolParams()->getBlock(block-1).comps == 0) {
+        for (size_t i = 1; i <= sim.getPatch(pidx)->getNoFields(basis); ++i)
+          components.insert(i);
+      } else
+        components = utl::getDigits(sim.getSolParams()->getBlock(block-1).comps);
+
+      for (size_t i = 0; i < lNodes[basis-1].size(); ++i) {
+        int node = lNodes[basis-1][i];
+        for (const int& dof : components) {
+          int eq = sim.getSAM()->getEquation(node, dof);
+          if (eq > 0) {
+            if (block == 0)
+              result.push_back(blocks[block].MLGEQ[eq-1]);
+            else
+              result.push_back(blocks[block].MLGEQ[blocks[block].G2LEQ[eq]-1]);
+          } else
+            result.push_back(0);
+        }
+      }
+    }
+  }
+
+  return result;
+}
+
+
+bool DomainDecomposition::calcGlobalEqNumbers(const ProcessAdm& adm,
+                                              const SIMbase& sim)
+{
+  // defaults in serial and non-parallel runs with MPI enabled builds.
+  blocks[0].minEq = 1;
+  blocks[0].maxEq = sim.getSAM()->getNoEquations();
+  for (size_t i = 1; i < blocks.size(); ++i) {
+    blocks[i].minEq = 1;
+    blocks[i].maxEq = blocks[i].localEqs.size();
+  }
+
+#ifdef HAVE_MPI
+  if (adm.getNoProcs() == 1)
+    return true;
+
+  IFEM::cout << "\tEstablishing global equation numbers" << std::endl;
+
+  auto locLMs = sim.getPatch(1)->getLagrangeMultipliers();
+  std::vector<int> glbLMs;
```

Comment:
Ja, poenget er at 'L' multipliers er lokale til en patch (eller element mer sannsynlig, men uansett) sånn at jeg ikke trenger å bry meg om de i parallell - de er som enhver intern node som ikke addresseres på andre prosesser.

Å flytte ting til SIM-nivå kompliserer vel neppe ting veldig mye som sådan, det ville vel gjøre ting enklere for denne koden faktisk.. Det sitter der det sitter fordi jeg tok utgangspunkt i hva som var der ('L' multipliers) og utvidet til det jeg trengte for å få timo opp og gå med cfd-kjøringene fortest mulig, ikke nødvendigvis fordi det er det beste valget.

Samtidig er det vel et litt 'alien' konsept at SIM'en interakterer mutable med SAM minus constructor / init-kall. Og multiplikatoren må uansett behandles likt nede i ASMbase - legges til alle element, noden må allokeres i SAM etc. Så jeg tror ikke det er mye å vinne på det annet enn den 'logiske godbiten', altså at SIM tar seg av ting som kun er for en kolleksjon av patcher. Men da vi kan ha globale multipliers på single-patch modeller okke som så blir jo det poenget  litt diffust.